### DISCUSSION TITLE
Explicit set `__NodeId` property in index

### DISCUSSION THREAD

**Discussion by bjarnef** [umbraco/Umbraco-CMS#13956] [Features and ideas]
It seems Umbraco implementation the `__NodeId` property is set based on `id` value passed into `ValueSet` ctor, which is confusing in a custom index.

E.g.

```
var course = model.Course;

var values = new Dictionary<string, IEnumerable<object>>()
{
	{ Constants.Examine.Common.FieldNames.Reference, new object[] { model.DocId }},
	{ Constants.Examine.FieldNames.HideFromSearch, NoValue },
	{ Umbraco.Cms.Infrastructure.Examine.UmbracoExamineFieldNames.ItemIdFieldName, new object[] { course.NodeId! }},
	{ Umbraco.Cms.Infrastructure.Examine.UmbracoExamineFieldNames.NodeNameFieldName, new object[] { course.Name! }},
	{ Constants.Examine.Course.FieldNames.CourseName, new object[] { course.Name! }},
	{ Constants.Examine.Course.FieldNames.CategoryId, new object[] { string.Join(" ", course.CategoryIds!) }},
	{ Constants.Examine.Course.FieldNames.BusinessId, new object[] { string.Join(" ", course.BusinessIds!) }},
};

var valueSet = new ValueSet(model.DocId, _category, _itemType, values);
```

![image](https://user-images.githubusercontent.com/2919859/224738838-50af160d-d829-46d5-a4d6-a12aba419ae8.png)

when we also included `Umbraco.Cms.Infrastructure.Examine.UmbracoExamineFieldNames.ItemIdFieldName` (`__NodeId`) the indexed document shows this, which seems to be `id` combined with `__NodeId` :

![image](https://user-images.githubusercontent.com/2919859/224740061-cc7446e7-27ff-469e-bad2-78bd3a07e059.png)

using this instead shows only this id:

```
var values = new Dictionary<string, IEnumerable<object>>()
{
	{ Constants.Examine.Common.FieldNames.Reference, new object[] { model.DocId }},
	{ Constants.Examine.FieldNames.HideFromSearch, NoValue },
	{ Umbraco.Cms.Infrastructure.Examine.UmbracoExamineFieldNames.NodeNameFieldName, new object[] { course.Name! }},
	{ Constants.Examine.Course.FieldNames.CourseName, new object[] { course.Name! }},
	{ Constants.Examine.Course.FieldNames.CategoryId, new object[] { string.Join(" ", course.CategoryIds!) }},
	{ Constants.Examine.Course.FieldNames.BusinessId, new object[] { string.Join(" ", course.BusinessIds!) }},
};

var valueSet = new ValueSet(model.DocId, _category, _itemType, values);
```

![image](https://user-images.githubusercontent.com/2919859/224740594-7b4ee72e-ffd5-44a2-9cd1-cd9950eb8656.png)

However still a bit confusing since `__NodeId` property isn't set directly in value set builder.

We can index this instead (or where value of constamt `Constants.Examine.Course.FieldNames.NodeId` is `id` or `courseId`).
It seems Umbraco use `id` in the content valueset builder: https://github.com/umbraco/Umbraco-CMS/blob/33adbf41fa1f5c5d0759c70a7116114107addf56/src/Umbraco.Infrastructure/Examine/ContentValueSetBuilder.cs#L80

```
var values = new Dictionary<string, IEnumerable<object>>()
{
	{ Constants.Examine.Common.FieldNames.Reference, new object[] { model.DocId }},
	{ Constants.Examine.FieldNames.HideFromSearch, NoValue },
	{ Umbraco.Cms.Infrastructure.Examine.UmbracoExamineFieldNames.NodeNameFieldName, new object[] { course.Name! }},
	{ Constants.Examine.Course.FieldNames.NodeId, new object[] { course.NodeId! }},
	{ Constants.Examine.Course.FieldNames.CourseName, new object[] { course.Name! }},
	{ Constants.Examine.Course.FieldNames.CategoryId, new object[] { string.Join(" ", course.CategoryIds!) }},
	{ Constants.Examine.Course.FieldNames.BusinessId, new object[] { string.Join(" ", course.BusinessIds!) }},
};

var valueSet = new ValueSet(model.DocId, _category, _itemType, values);
```

Not sure if some logic by default set `__NodeTypeAlias` and `__NodeId` even when it is a custom implementation of `IValueSetBuilder`.

**Comment by bjarnef** [comment]
@Zeegaan @kjac do you have any idea why this happens? I would have expected it just just `__NodeId` property and the `id` passed into `ValueSet` is the document id?

Is there any reason it always set `__NodeId` and `__NodeTypeAlias` in a custom index?

**Reply by bergmania** [reply]
`__NodeId`, `__NodeTypeAlias` and `__IndexType` are constants in Examine too. I think the `LuceneIndex` which we inherit from sets it.
